src/lib.rs

```diff
@@ -112,129 +155,134 @@ pub fn glob_with(pattern: &str, options: &MatchOptions) -> Paths {
 
-                // no non-recursive patterns follow the current one
-                // so auto-accept all remaining recursive paths
-                if next == self.dir_patterns.len() {
+                // the path is a directory, so it's a match
+                if path.is_dir() {
+                    // push this directory's contents
                     fill_todo(&mut self.todo, self.dir_patterns.as_slice(),
-                              next - 1, &path, &self.options);
-                    return Some(path);
-                }
+                              next, &path, &self.options);
 
-                let ref next_pattern = self.dir_patterns[next];
-                let is_match = next_pattern.matches_with(match path.filename_str() {
-                    // this ugly match needs to go here to avoid a borrowck error
-                    None => {
-                        // FIXME (#9639): How do we handle non-utf8 filenames? Ignore them for now
-                        // Ideally we'd still match them against a *
-                        continue;
+                    // pattern ends in recursive pattern, so return this directory as a result
+                    if next == self.dir_patterns.len() - 1 {
+                        return Some(Ok(path));
                     }
-                    Some(x) => x
-                }, &self.options);
-
-                // determine how to advance
-                let (current_idx, next_idx) =
-                    if is_match {
-                        // accept the pattern after the next non-recursive pattern
-                        (next, next + 1)
-                    } else {
-                        // next pattern still hasn't matched
-                        // so stay on this recursive pattern
-                        (next - 1, next - 1)
-                    };
-
-                if current_idx == self.dir_patterns.len() - 1 {
-                    // it is not possible for a pattern to match a directory *AND* its children
-                    // so we don't need to check the children
 
-                    if !self.require_dir || path.is_dir() {
-                        return Some(path);
+                    // advanced to the next pattern for this path
+                    else {
+                        idx = next + 1;
                     }
-                } else {
-                    fill_todo(&mut self.todo, self.dir_patterns.as_slice(),
-                              next_idx, &path, &self.options);
                 }
-            }
 
-            // it's recursive and it's the last pattern
-            // automatically match everything else recursively
-            else if is_recursive && is_last {
-              fill_todo(&mut self.todo, self.dir_patterns.as_slice(),
-                        idx, &path, &self.options);
-              return Some(path);
+                // advanced to the next pattern for this path
+                else if next != self.dir_patterns.len() - 1 {
```

Comment:
I think the typical style for else-if is to have the closing brace of the previous if on the same line as the else.